### PR TITLE
Rms 2 candidate crud

### PR DESCRIPTION
**Pull request by lra-wtag**
[created candidate crud](https://welltravel.atlassian.net/browse/RMS-2)

### CODE REVIEW

File: app/views/candidates/new.html.erb [hunk 1]
```diff
@@ -0,0 +1,7 @@
+<%= link_to t('shared.all.can_back'), candidates_path %>
+<div class ="candidate new">
+  <h2><%= t('candidate.can_new') %></h2>
+  <%= simple_form_for @candidate do |f| %>
+  <%= render(partial: 'form', locals: {f: f}) %>
```
Comment (fnu-wtag) [inline app/views/candidates/new.html.erb]: please follow previous instruction

File: bin/setup [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+chdir APP_ROOT do
+  # This script is a starting point to setup your application.
+  # Add necessary setup steps to this file.
+
+  puts '== Installing dependencies =='
+  system! 'gem install bundler --conservative'
+  system('bundle check') || system!('bundle install')
+
+  # Install JavaScript dependencies if using Yarn
+  # system('bin/yarn')
```
Comment (fnu-wtag) [inline bin/setup]: This file changes were not supposed to be added here because of 'gitignore'

File: Gemfile [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+gem 'pg'
+gem 'rails_12factor', group: :production
+gem 'dotenv-rails'
+gem 'puma'
+gem 'sass-rails'
+gem 'uglifier'
+gem 'coffee-rails'
+gem 'turbolinks'
+gem 'jbuilder'
+gem 'bootsnap', require: false
```
Comment (sde-wtag) [inline Gemfile]: Please use `gem 'bootstrap-sass'` here. `bootsnap` Typo? 😲

File: Gemfile [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
+gem 'tzinfo-data', platforms: [:mingw, :mswin, :x64_mingw, :jruby]
+gem 'rails-controller-testing'
+
+group :development, :test do
+  gem 'byebug', platforms: [:mri, :mingw, :x64_mingw]
+end
+
+group :development, :test do
+  gem "rspec-rails"
```
Comment (sde-wtag) [inline Gemfile]: May be you have used this group two times. Please keep it in one group.

File: Gemfile [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+  gem "factory_girl_rails"
```
Comment (sde-wtag) [inline Gemfile]: Please use `gem "factory_bot_rails"` here.

File: Gemfile [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+  gem 'spring-watcher-listen'
+end
+
+group :test do
+  gem 'faker'
+  gem 'capybara'
+  gem 'database_cleaner'
+  gem 'launchy'
+  gem 'selenium-webdriver'
+  gem 'chromedriver-helper'
```
Comment (sde-wtag) [inline Gemfile]: May be no need to use `capybara`, 'selenium-webdriver', 'chromedriver-helper' and `launchy` gem.

File: app/controllers/application_controller.rb [hunk 7]
```diff
@@ -0,0 +1,7 @@
+class ApplicationController < ActionController::Base
+  before_action :set_locale
+  def set_locale
+    locale = params[:locale].to_s.strip.to_sym
```
Comment (sde-wtag) [inline app/controllers/application_controller.rb]: params[:locale] may be comes as a string. So need to_s here.

File: app/models/candidate.rb [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+  validates :contact, numericality: true
+
+  before_validation { avatar.delete if @delete_avatar }
+
+  def delete_avatar
+    @delete_avatar ||= false
+  end
+
+  def delete_avatar=(value)
+    @delete_avatar  = !value.to_i.zero?
```
Comment (sde-wtag) [inline app/models/candidate.rb]: Please try to avoid negative sign. May be you can use value.to_i.nonzero?

File: app/models/candidate.rb [hunk 9]
```diff
@@ -0,0 +1,21 @@
+class Candidate < ApplicationRecord
+  has_one_attached :avatar
+  Gender = %w[Male Female]
```
Comment (sde-wtag) [inline app/models/candidate.rb]: For constant variable, please use block letters. And use %w(Male Female). No need to use this `[]`. And please also apply wherever you have used this sign.

File: app/assets/stylesheets/admin_additions.css.scss [hunk 10]
```diff
@@ -0,0 +1,116 @@
+#content {
```
Comment (fnu-wtag) [inline app/assets/stylesheets/admin_additions.css.scss]: all stylesheets will be only `.scss` file, no need to add `.css`

File: app/channels/application_cable/channel.rb [hunk 11]
```diff
@@ -0,0 +1,4 @@
+module ApplicationCable
```
Comment (fnu-wtag) [inline app/channels/application_cable/channel.rb]: if not needed then get rid of this files

File: app/controllers/candidates_controller.rb [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+  def new
+    @candidate = Candidate.new
+  end
+
+  def create
+    @candidate = Candidate.new(candidate_params)
+    @candidate.avatar.attach(candidate_params[:avatar])
+    if @candidate.save
+      flash[:notice] = t('candidate.can_notice_create', candidate_name: @candidate.name)
+      redirect_to(candidates_path)
```
Comment (fnu-wtag) [inline app/controllers/candidates_controller.rb]: remove () brackets

File: app/controllers/candidates_controller.rb [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+      render :new
+    end
+  end
+
+  def edit ; end
+
+  def update
+    if @candidate.update_attributes(candidate_params)
+      flash[:notice] = t('candidate.can_notice_edit', candidate_name: @candidate.name)
+      redirect_to(candidates_path(@candidate))
```
Comment (fnu-wtag) [inline app/controllers/candidates_controller.rb]: remove () brackets, it's not necessary 
`redirect_to candidates_path(@candidate)`

File: app/controllers/candidates_controller.rb [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+      render :edit
+    end
+  end
+
+  def delete ; end
+
+  def destroy
+    @candidate.destroy
+    flash[:notice] = t('candidate.can_notice_delete', candidate_name: @candidate.name)
+    redirect_to(candidates_path)
```
Comment (fnu-wtag) [inline app/controllers/candidates_controller.rb]: same as previous

File: app/helpers/application_helper.rb [hunk 15]
```diff
@@ -0,0 +1,2 @@
+module ApplicationHelper
```
Comment (fnu-wtag) [inline app/helpers/application_helper.rb]: As we didn't use any helper methods here, we can get rid of this files

File: app/helpers/candidates_helper.rb [hunk 16]
```diff
@@ -0,0 +1,2 @@
+module CandidatesHelper
```
Comment (fnu-wtag) [inline app/helpers/candidates_helper.rb]: same as above

File: app/jobs/application_job.rb [hunk 17]
```diff
@@ -0,0 +1,2 @@
+class ApplicationJob < ActiveJob::Base
```
Comment (fnu-wtag) [inline app/jobs/application_job.rb]: same as above

File: app/models/candidate.rb (hunk 9 above)
Comment (fnu-wtag) [inline app/models/candidate.rb]: We need to use i18n here

File: app/views/candidates/_form.html.erb [hunk 18]
```diff
@@ -0,0 +1,26 @@
+<%= f.error_notification %>
+<%= f.input :name, label: t('candidate.can_name') %>
+<%= f.input :gender, label: t('shared.all.gender'), as: :radio_buttons, collection: (Candidate::Gender), checked: 'Male'%>
```
Comment (fnu-wtag) [inline app/views/candidates/_form.html.erb]: use i18n

File: config/locales/en.yml [hunk 19]
```diff
@@ -0,0 +1,39 @@
+en:
+  shared:
+      all:
+        action: 'Actions'
```
Comment (fnu-wtag) [inline config/locales/en.yml]: we don't need to use single quotations, just write `action: Actions`

File: spec/models/candidate_spec.rb [hunk 20]
```diff
@@ -0,0 +1,13 @@
...
+  it { is_expected.to validate_presence_of(:name) }
+  it { is_expected.to validate_presence_of(:email) }
+  it { is_expected.to validate_presence_of(:skill) }
+  it { is_expected.to validate_presence_of(:experience) }
+  it { is_expected.to validate_presence_of(:personal_interest) }
+  it 'is invalid with a duplicate email' do
+    create(:candidate)
+    is_expected.to validate_uniqueness_of(:email)
+  end
+  it { should validate_numericality_of(:experience) }
```
Comment (fnu-wtag) [inline spec/models/candidate_spec.rb]: we use `is_expected.to` instead of should

File: spec/factories/candidates.rb [hunk 21]
```diff
@@ -0,0 +1,17 @@
...
+    contact { '01792050217' }
+    address { 'house 48, road 4, bashundhara' }
+    skill { 'c++,c,ruby' }
+    experience { '2' }
+    personal_interest { 'shopping' }
+    hobbies { 'travelling' }
+    long_term_plan { 'CEO' }
+    keywords { 'software' }
+    referrals { 'Tanzim' }
+    email {Faker::Internet.email}
```
Comment (sde-wtag) [inline spec/factories/candidates.rb]: We use, FFaker normally. So, if you didn't use this gem, then use FFaker gem.

File: spec/rails_helper.rb [hunk 22]
```diff
@@ -0,0 +1,61 @@
+# This file is copied to spec/ when you run 'rails generate rspec:install'
+ENV["RAILS_ENV"] ||= 'test'
+require 'spec_helper'
+require File.expand_path("../../config/environment", __FILE__)
+require 'rspec/rails'
+require 'factory_bot_rails'
+#require 'support/factory_bot'
```
Comment (sde-wtag) [inline spec/rails_helper.rb]: remove this comment

File: app/helpers/candidates_helper.rb [hunk 23]
```diff
@@ -0,0 +1,5 @@
...
+  def human_attribute_genders
+    Hash[Candidate.genders.map { |k,v| [k, Candidate.human_attribute_name("gender.#{k}")] }]
```
Comment (sde-wtag) [inline app/helpers/candidates_helper.rb]: Write a test for this.

File: app/controllers/application_controller.rb [hunk 24]
```diff
@@ -0,0 +1,6 @@
+class ApplicationController < ActionController::Base
+  before_action :set_locale
+  def set_locale
+    I18n.locale = params[:locale] || I18n.default_locale
```
Comment (sde-wtag) [inline app/controllers/application_controller.rb]: Write a test for this.

File: .byebug_history [hunk 25]
```diff
@@ -0,0 +1,9 @@
+@candidate
```
Comment (sde-wtag) [inline .byebug_history]: Please remove this file or add this into gitignore.

File: app/models/candidate.rb [hunk 26]
```diff
@@ -0,0 +1,24 @@
+class Candidate < ApplicationRecord
+  has_one_attached :avatar
+
+  enum gender: [:male, :female]
+
+  scope :sorted, lambda { order('experience ASC') }
```
Comment (sde-wtag) [inline app/models/candidate.rb]: Write a spec for this.

File: app/models/candidate.rb [hunk 27]
```diff
@@ -0,0 +1,24 @@
...
+
+  validates :name, presence: true
+  validates :email, presence: true, uniqueness: true,
+            format: { with: /\A[\w+\-.]+@[a-z\d\-]+(\.[a-z\d\-]+)*\.[a-z]+\z/i }
+  validates :skill,:personal_interest, presence: true
+  validates :experience, presence: true, numericality: true
+  validates :contact, numericality: true
+
+  before_validation { avatar.delete if @delete_avatar }
```
Comment (sde-wtag) [inline app/models/candidate.rb]: Write a test for this.

File: app/models/candidate.rb [hunk 28]
```diff
@@ -0,0 +1,24 @@
+class Candidate < ApplicationRecord
+  has_one_attached :avatar
+
+  enum gender: [:male, :female]
+
+  scope :sorted, lambda { order('experience ASC') }
+
+  validates :name, presence: true
+  validates :email, presence: true, uniqueness: true,
+            format: { with: /\A[\w+\-.]+@[a-z\d\-]+(\.[a-z\d\-]+)*\.[a-z]+\z/i }
```
Comment (sde-wtag) [inline app/models/candidate.rb]: Write a test for this format.

File: spec/models/candidate_spec.rb [hunk 29]
```diff
@@ -0,0 +1,30 @@
...
+  it { is_expected.to validate_presence_of(:experience) }
+  it { is_expected.to validate_presence_of(:personal_interest) }
+
+  it 'is invalid with a duplicate email' do
+    create(:candidate)
+    is_expected.to validate_uniqueness_of(:email)
+  end
+
+  it 'has the correct format' do
+    expect(subject).to allow_value(FFaker::Internet.email).for(:email)
```
Comment (sde-wtag) [inline spec/models/candidate_spec.rb]: Use real example for this test. and you can use `should-a-matchers` here.

File: spec/helper/candidates_helper_spec.rb [hunk 30]
```diff
@@ -0,0 +1,21 @@
...
+  describe '#human_attribute_genders' do
+    subject { human_attribute_genders }
+
+    context 'prepares hash value for gender' do
+      let(:candidate) do
+        FactoryBot.create(:candidate)
+      end
+      let(:gender) do
+        {
+            'female' => 'FEMALE',
```
Comment (sde-wtag) [inline spec/helper/candidates_helper_spec.rb]: indentation

File: app/assets/stylesheets/admin.scss [hunk 31]
```diff
@@ -0,0 +1,53 @@
+/*
+ *= require_tree .
+ *= require_self
+ */
+.simple_form label {
+  display: inline-block;
+  width: 875px;
+  text-align: right;
+  margin: 10px;
+
```
Comment (mal-wtag) [inline app/assets/stylesheets/admin.scss]: Can we remove this extra line ?

File: app/assets/stylesheets/primary.scss [hunk 32]
```diff
@@ -0,0 +1,149 @@
+$red: #993120;
+$green: #94B26F;
+$light_green: #B6D491;
+$yellow: #F2DC87;
+$dark_brown: #332318;
+$light_brown: #4C2516;
+$white: #FFFFFF;
```
Comment (mal-wtag) [inline app/assets/stylesheets/primary.scss]: Is these variables are being used anywhere? if so how about we write them in a `variable.scss` for future good tracking.

File: app/assets/stylesheets/primary.scss [hunk 33]
```diff
@@ -0,0 +1,149 @@
...
+
+.clear { clear: both; }
+.center { text-align: center; }
+
+.alert.alert-failure.alert-note.main-alert{
+  text-align: center;
+  position: relative;
+  height: 1.25em;
+  padding: 0.5em;
+  background: cornflowerblue;
```
Comment (mal-wtag) [inline app/assets/stylesheets/primary.scss]: We usually use hash or rgba , not word colorings. Maybe consult with @mmo-wtag  vhai about this

**Pull request merged**